File: rust/src/leetcode/Q739_daily_temperatures.rs
/*
 * @lc app=leetcode id=739 lang=rust
 *
 * [739] Daily Temperatures
 */

// @lc code=start
impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut stack: Vec<usize> = vec![];
        let mut result: Vec<i32> = vec![0; temperatures.len()];

        for i in 0..temperatures.len() {
            while let Some(&previous_index) = stack.last() {
                if temperatures[i] > temperatures[previous_index] {
                    stack.pop();
                    result[previous_index] = (i - previous_index) as i32;
                } else {
                    break;
                }
            }

            stack.push(i);
        }

        result
    }
}
// @lc code=end

File: python/Q739_daily_temperatures.py
#
# @lc app=leetcode id=739 lang=python3
#
# [739] Daily Temperatures
#


# @lc code=start
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        stack: list[int] = []
        result: list[int] = [0] * len(temperatures)

        for i, t in enumerate(temperatures):
            while stack and t > temperatures[stack[-1]]:
                index = stack.pop()
                result[index] = i - index

            stack.append(i)

        return result


# @lc code=end
